File: data_structure/linked_list/singly_linked_list.py
"""
Linked List is a linear data structure. Unlike arrays, linked list elements
are not stored at a contiguous location the elements are linked using
pointers. Singly linked list is represented by a pointer to the first node
called head, if linked list is empty then the value of head is NULL (None).
Each node of a singly linked list contains 2 parts:
  - Stored data
  - Pointer to the next node (or reference)
      HEAD
    [data1|-]->[data2|-]->[data3|-]->[data4|-]->[data5|-]->NULL
"""
from __future__ import annotations

import unittest
from typing import TypeVar, Generic, Generator, Any

T = TypeVar("T")


class SinglyLinkedList(Generic[T]):
    class Node:
        def __init__(self, data: T, nxt: Node = None):
            self._data = data
            self._nxt = nxt

        def __repr__(self) -> str:
            return f"[{self._data}]->[]" if self._nxt is None else f"[{self._data}]->"

        @property
        def data(self) -> T:
            return self._data

        @data.setter
        def data(self, data):
            self._data = data

        @property
        def nxt(self) -> Node:
            return self._nxt

        @nxt.setter
        def nxt(self, nxt):
            self._nxt = nxt

    def __init__(self, head: Node = None, size: int = 0):
        self._head = head
        self._size = size

    def __iter__(self) -> Generator[Node, Any, None]:
        tmp = self._head
        while tmp:
            yield tmp
            tmp = tmp.nxt

    def __repr__(self) -> str:
        return "".join([str(node) for node in self])

    @property
    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._head is None

    def search(self, data: T) -> bool:
        tmp = self._head
        while tmp:
            if tmp.data == data:
                return True
            tmp = tmp.nxt
        return False

    def insert(self, data: T):
        node = SinglyLinkedList.Node(data)
        if self.is_empty():
            self._head = node
        else:
            tmp = self._head
            while tmp.nxt:
                tmp = tmp.nxt
            tmp.nxt = node
        self._size += 1

    def remove(self, data: T) -> T:
        if self.is_empty():
            raise RuntimeError("List is empty, can not remove")

        old, prev = self._head, None
        while old and old.data != data:
            prev = old
            old = old.nxt

        if not old:
            raise RuntimeError("Node is not exist in this list")

        removed = old.data
        if old is self._head:
            self._head = self._head.nxt
        else:
            prev.nxt = old.nxt

        self._size -= 1
        return removed


class TestSinglyLinkedList(unittest.TestCase):
    def test_integer_linked_list(self):
        linked_list: SinglyLinkedList[int] = SinglyLinkedList()

        self.assertTrue(linked_list.is_empty())
        self.assertEqual(linked_list.size, 0)

        linked_list.insert(8)
        linked_list.insert(12)
        linked_list.insert(6)
        linked_list.insert(10)

        self.assertFalse(linked_list.is_empty())
        self.assertListEqual([node.data for node in linked_list], [8, 12, 6, 10])

        self.assertTrue(linked_list.search(8))
        self.assertFalse(linked_list.search(9))

        linked_list.remove(10)
        self.assertListEqual([node.data for node in linked_list], [8, 12, 6])
        linked_list.remove(6)
        self.assertListEqual([node.data for node in linked_list], [8, 12])

        with self.assertRaises(RuntimeError):
            linked_list.remove(9)

        linked_list.remove(8)
        self.assertListEqual([node.data for node in linked_list], [12])
        linked_list.remove(12)
        self.assertListEqual([node.data for node in linked_list], [])

        with self.assertRaises(RuntimeError):
            linked_list.remove(9)


if __name__ == "__main__":
    unittest.main()

File: data_structure/linked_list/singly_linked_list.cc
#include "singly_linked_list.h"

#include "utils/test.h"

namespace data_structure {
namespace linked_list {

template <typename T>
SinglyLinkedList<T>::SinglyLinkedList() : head_(NULL), size_(0) {}

template <typename T>
SinglyLinkedList<T>::~SinglyLinkedList() {
  while (!IsEmpty()) {
    Remove(head_->data_);
  }
  delete head_;
}

template <typename T>
std::size_t SinglyLinkedList<T>::Size() const {
  return static_cast<std::size_t>(size_);
}

template <typename T>
bool SinglyLinkedList<T>::IsEmpty() const {
  return (head_ == NULL);
}

template <typename T>
bool SinglyLinkedList<T>::Search(const T &data) const {
  auto *temp = head_;
  while (temp != NULL) {
    if (temp->data_ == data) {
      return true;
    }
    temp = temp->next_;
  }
  return false;
}

template <typename T>
void SinglyLinkedList<T>::Insert(const T &data) {
  Node *node = new Node(data);
  if (IsEmpty()) {
    head_ = node;
  } else {
    Node *temp = head_;
    while (temp->next_ != NULL) {
      temp = temp->next_;
    }
    temp->next_ = node;
  }
  size_++;
}

template <typename T>
T SinglyLinkedList<T>::Remove(const T &data) {
  if (IsEmpty()) {
    throw std::runtime_error("List is empty, can not remove");
  }

  Node *old = head_, *prev;
  while (old != NULL && old->data_ != data) {
    prev = old;
    old = old->next_;
  }

  if (old == NULL) {
    throw std::runtime_error("Node is not exist in this list");
  }

  auto removed = old->data_;
  if (old == head_) {
    head_ = head_->next_;
  } else {
    prev->next_ = old->next_;
  }

  delete old;
  size_--;
  return removed;
}

namespace test {

void SinglyLinkedList_FunctionalTest() {
  SinglyLinkedList<int32_t> linked_list;

  EXPECT_EQ(linked_list.IsEmpty(), true);
  EXPECT_EQ(linked_list.Size(), 0);

  linked_list.Insert(8);
  linked_list.Insert(12);
  linked_list.Insert(6);
  linked_list.Insert(10);

  EXPECT_EQ(linked_list.Size(), 4);
  EXPECT_EQ(linked_list.IsEmpty(), false);
  std::cout << linked_list << std::endl;

  EXPECT_EQ(linked_list.Search(8), true);
  EXPECT_EQ(linked_list.Search(9), false);

  EXPECT_EQ(linked_list.Remove(10), 10);
  EXPECT_EQ(linked_list.Size(), 3);

  EXPECT_EQ(linked_list.Remove(6), 6);
  EXPECT_EQ(linked_list.Size(), 2);

  EXPECT_THROW(linked_list.Remove(6));
}

void RunTests() { TEST(SinglyLinkedList_FunctionalTest); }

}  // namespace test

}  // namespace linked_list
}  // namespace data_structure

int main() {
  std::cout << "-----Running Singly Linked List tests-----" << std::endl;
  data_structure::linked_list::test::RunTests();
  return 0;
}